dmoj/cptbox/isolate.py

```diff
@@ -201,176 +202,290 @@ def __init__(self, read_fs, write_fs=None, writable=(1, 2)):
                     sys_minherit: ALLOW,
                     sys_thr_set_name: ALLOW,
                     sys_sigfastblock: ALLOW,
-                    sys_realpathat: self.check_file_access_at('realpathat', FilesystemSyscallKind.READ),
+                    sys_realpathat: self.handle_file_access_at('realpathat', FilesystemSyscallKind.READ),
                 }
             )
 
-    def _compile_fs_jail(self, fs):
+    def _compile_fs_jail(self, fs) -> FilesystemPolicy:
         return FilesystemPolicy(fs or [])
 
-    @staticmethod
-    def read_path(syscall: str, debugger: Debugger, ptr: int):
-        try:
-            file = debugger.readstr(ptr)
-        except MaxLengthExceeded as e:
-            log.warning('Denied access via syscall %s to overly long path: %r', syscall, e.args[0])
-            return None, ACCESS_ENAMETOOLONG(debugger)
-        except UnicodeDecodeError as e:
-            log.warning('Denied access via syscall %s to path with invalid unicode: %r', syscall, e.object)
-            return None, ACCESS_ENOENT(debugger)
-        return file, None
-
-    def _fs_jail_from_open_flags(self, open_flags: int) -> bool:
-        for flag in open_write_flags:
-            # Strict equality is necessary here, since e.g. O_TMPFILE has multiple bits set,
-            # and O_DIRECTORY & O_TMPFILE > 0.
-            if open_flags & flag == flag:
-                return self.write_fs_jail
-
-        return self.read_fs_jail
-
-    def _fs_jail(self, read_open_flags, kind):
-        if kind == FilesystemSyscallKind.WRITE:
-            return lambda _: self.write_fs_jail
-        elif kind in (FilesystemSyscallKind.READ, FilesystemSyscallKind.FSTAT):
-            return lambda _: self.read_fs_jail
-        elif kind == FilesystemSyscallKind.OPEN:
-            return lambda debugger: self._fs_jail_from_open_flags(read_open_flags(debugger))
-        else:
-            raise ValueError('bad FilesystemSyscallKind value')
-
-    def check_file_access(self, syscall, argument, kind) -> HandlerCallback:
-        fs_jail = self._fs_jail(attrgetter('uarg1'), kind)
-
-        def check(debugger: Debugger) -> bool:
-            file, error = self.read_path(syscall, debugger, getattr(debugger, 'uarg%d' % argument))
-            if error is not None:
-                return error
-
-            file, error = self._file_access_check(file, debugger, fs_jail(debugger))
-            if not error:
-                return True
+    def handle_file_access(self, syscall: str, kind: FilesystemSyscallKind, file_reg: int = 0) -> HandlerCallback:
```

Comment:
```suggestion
    def handle_file_access(self, syscall: str, *, kind: FilesystemSyscallKind, file_reg: int = 0) -> HandlerCallback:
```

and fix resulting errors